src/commands/handlers/registration.js

```diff
@@ -213,7 +213,11 @@ const handlers = {
                     ) {
                         handler.connection.write('AUTHENTICATE ' + wanted_mechanism);
                         authenticating = true;
+                    } else {
+                        handleSaslFail(handler, 'unsupported_mechanism');
                     }
+                } else if (handler.connection.options.account) {
+                    handleSaslFail(handler, 'capability_missing');
```

Comment:
Here too... by now you need to keep track a bunch of bools in your head if you want to follow the logic.
Not sure if it would be more readable if you start by invalidating all the invalid combination at the beginning but it would keep the happy path somewhat clearer, maybe.

Reply:
Also, are we ever sending CAP end here?
Same with the other error cases that aren't ERR_SASLFAIL we aren't emitting that at least in the code visible in this PR.
Do I misunderstand the spec?

Reply:
handleSaslFail() emits 'sasl failed' event for every call and handles the disconnection if requested,

this else if is really !enabled.sasl && account is set (meaning the client wants to sasl but the server does not support it)

I will add some more comments around this code to hopefully make it clearer